File: C++/mergesort.cc
//MERGE and MERGESORT
template <typename T>
void merge(T *A, int low, int mid, int high)
{
	int i, j, k;
	int n1 = mid - low + 1;
	int n2 = high - mid;
	
	//temp arrays left and right
	T* L = new T[n1];
	T* R = new T[n2];
	
	//Copy data temp arrays L[] and R[]
	for(i = 0; i < n1; i++)
		L[i] = A[low+i];
	for(j = 0; j < n2; j++)
		R[j] = A[mid+1+j];
	
	//Merge the temp arrays back into A[low..high]
	i = 0; //Initial index of first sub-array
	j = 0; //Initial index of second sub-array
	k = low; //Initial index of merged sub-array
	
	while(i < n1 && j < n2)
	{
		if(L[i] <= R[j])
		{
			A[k] = L[i];
			i++;
		}
		else 
		{
			A[k] = R[j];
			j++;
		}
		k++;
	}
	
	//Copy the remaining elements of L[], if there are any
	while(i < n1)
	{
		A[k] = L[i];
		i++; 
		k++;
	}
	
	//Copy the remaining elements of R[], if there are any
	while(j < n2)
	{
		A[k] = R[j];
		j++;
		k++;
	}
	
	delete[] R;
	delete[] L;
};

template <typename T>
void MergeSort(T *A, int low, int high)
{
	int mid;
	
	//Check for base case
	if(low < high)
	{
		//First, divide in half
		mid = low + (high - low)/2;
		
		//First recursive call
		MergeSort(A, low, mid);
		//Second recursive call
		MergeSort(A, mid+1, high);
		//The merge function call
		merge(A, low, mid, high);
	}
};

File: Python/mergesort.py
def merge(L, R, arr):
    '''Merge two sorted lists L and R into properly sized list arr'''
    i=j=0
    while i+j < len(arr):
        if j == len(R) or (i < len(L) and L[i] < R[j]):
                arr[i+j] = L[i]     # copy ith element of L as next item of arr
                i += 1
        else:
            arr[i+j] = R[j]         # copy jth element of R as next item of arr
            j += 1

def merge_sort(arr):
    ''' Sor the elements of list arr'''
    n = len(arr)
    if n < 2:
        return
    mid = n//2
    L = arr[0:mid]  #first half of list - Left side
    R = arr[mid:n]  #second half of list - Right side
    merge_sort(L)   #sort first half
    merge_sort(R)   #sort second half
    merge(L, R, arr)#merge halfs back together

